### PR TITLE
First website prototype

### PR DESCRIPTION
**Pull request by manningwu07**
Protected routes, changing user membership status, and AP Links (not built yet but linked)

### CODE REVIEW

**Comment by manningwu07** [comment]
Website prototype is done :O

(Admin stuff, User stuff, Firebase, etc.)

Only need styling and code efficency configurations now. + some other minor details like the git diff stuff.

File: src/app/admin/page.tsx [hunk 1]
```diff
@@ -1,60 +1,251 @@
...
 import Footer from "@/components/ui/footer";
-import APLibrary from "@/components/landingPage/APLibrary";
-import { getUser } from "@/components/hooks/getUser";
+import { getUser } from "@/components/hooks/users";
 import { User } from "@/types/user";
 import { useState, useEffect } from "react";
 import { useRouter } from "next/navigation";
+import { useUserManagement } from "./useUserManagement";
+
+const apClasses = [
```
Comment (alexng353) [inline src/app/admin/page.tsx:10]: nit- can you move this to an `apClasses.json` somewhere adjacent to this file for both re-use and conceptual isolation? This shouldn't be at the top of the file because incoming devs and maintainers shouldn't be reading it every time they open this file.

File: src/app/admin/page.tsx [hunk 2]
```diff
@@ -1,60 +1,251 @@
...
+        router.push("/");
       } finally {
-        setIsLoading(false);
+        setIsLoadingUser(false);
       }
     };
 
     fetchUser();
-  }, []);
+  }, [router]);
```
Comment (alexng353) [inline src/app/admin/page.tsx:90]: Shouldn't this kind of logic be in a global context...

File: src/app/admin/page.tsx [hunk 3]
```diff
@@ -1,60 +1,251 @@
...
+  "AP World History: Modern",
+].sort();
 
 const Page = () => {
-  const [user, setUser] = useState<User | null>(null);
-  const [isLoading, setIsLoading] = useState(true);
+  const [state, setState] = useState<{
+    user: User | null;
+    searchTermAPClasses: string;
+    searchTermUsers: string;
```
Comment (alexng353) [inline src/app/admin/page.tsx:56]: separate these into their own states

File: src/app/admin/page.tsx [hunk 4]
```diff
@@ -1,60 +1,251 @@
...
+                    value={state.searchTermUsers}
+                    onChange={(e) =>
+                      setState((prevState) => ({
+                        ...prevState,
+                        searchTermUsers: e.target.value,
+                      }))
+                    }
+                  />
+                  <ul className="class-list max-h-60 overflow-y-auto">
+                    {users!.map((u) => (
```
Comment (alexng353) [inline src/app/admin/page.tsx:147]: questionable usage of non-null coercion, leave a comment explaining how and why `users` will never be nullable

File: src/app/admin/page.tsx [hunk 5]
```diff
@@ -1,60 +1,251 @@
...
+                      <li
+                        key={u.uid}
+                        className="grid cursor-pointer grid-cols-1 gap-4 rounded-md border p-4 text-center shadow-md hover:bg-gray-200 md:grid-cols-2 lg:grid-cols-3"
+                        onClick={() => openModal(u)} // Open modal on click
```
Comment (alexng353) [inline src/app/admin/page.tsx:151]: this comment does *not* need to exist

File: src/app/admin/page.tsx [hunk 6]
```diff
@@ -1,60 +1,251 @@
...
+                  <li
+                    key={apClass}
+                    className="cursor-pointer rounded-md p-2 hover:bg-gray-200"
+                    onClick={() =>
+                      router.push(
+                        `/admin/subject/${apClass
+                          .replace(/AP /g, "")
+                          .toLowerCase()
+                          .replace(/[^a-z1-9 ]+/g, "")
+                          .replace(/\s/g, "-")}`,
```
Comment (alexng353) [inline src/app/admin/page.tsx:196]: Ideally this regex should be in a function named `formatAPClassName` (or something) so it can be reused easily

File: src/app/admin/subject/[slug]/[unit]/[articleNumber]/page.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import { User } from "@/types/user";
+import { useEffect, useState } from "react";
+
+const Page = ({ params }: { params: { slug: string } }) => {
+  const [loading, setLoading] = useState<boolean>(true);
+  const [error, setError] = useState<string | null>(null);
+  const [user, setUser] = useState<User | null>(null);
+
+  const pathParts = window.location.pathname.split("/").slice(-2);
+  const formattedTitle = `Unit ${pathParts[1]?.charAt(0).toUpperCase() + pathParts[1]!.slice(1)}: ${pathParts[0]?.charAt(0).toUpperCase() + pathParts[0]!.slice(1)}`.replace(/-/g, " ");
```
Comment (alexng353) [inline src/app/admin/subject/[slug]/[unit]/[articleNumber]/page.tsx:15]: ???

File: src/app/admin/subject/[slug]/[unit]/[articleNumber]/page.tsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import Navbar from "@/components/ui/navbar";
+import { User } from "@/types/user";
+import { useEffect, useState } from "react";
+
+const Page = ({ params }: { params: { slug: string } }) => {
+  const [loading, setLoading] = useState<boolean>(true);
+  const [error, setError] = useState<string | null>(null);
+  const [user, setUser] = useState<User | null>(null);
+
+  const pathParts = window.location.pathname.split("/").slice(-2);
```
Comment (alexng353) [inline src/app/admin/subject/[slug]/[unit]/[articleNumber]/page.tsx:14]: Get this from nextRouter instead of using window

File: src/app/admin/subject/[slug]/page.tsx [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+  "AP Psychology",
+  "AP Research",
+  "AP Seminar",
+  "AP Spanish Language",
+  "AP Spanish Literature",
+  "AP Statistics",
+  "AP US History",
+  "AP United States Government",
+  "AP World History: Modern",
+].sort();
```
Comment (alexng353) [inline src/app/admin/subject/[slug]/page.tsx:53]: Case in point, this should be in some central file so you don't need to copypaste it

File: src/app/admin/subject/[slug]/page.tsx [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+      try {
+        if (user && (user?.access === "admin" || user?.access === "member")) {
+          // Reference to the document in Firestore using the slug
+          const docRef = doc(db, "subjects", params.slug);
+          const docSnap = await getDoc(docRef);
+          if (docSnap.exists()) {
+            // Convert Firestore document data to Subject type
+            setSubject(docSnap.data() as Subject);
+          } else {
+            emptyData.title = apClasses.find((apClass) => apClass.replace(/AP /g, "").toLowerCase().replace(/[^a-z1-9 ]+/g, "").replace(/\s/g, "-") === params.slug) || "";
```
Comment (alexng353) [inline src/app/admin/subject/[slug]/page.tsx:102]: ???

Can you use a "formatAPClassname" function or something at this point

File: src/app/admin/subject/[slug]/page.tsx [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    try {
+      if (user && (user?.access === "admin" || user?.access === "member")) {
+
+        // Save to Firestore
+        await setDoc(doc(db, "subjects", params.slug), subject);
+
+        setSubject(subject);
+      } else {
+        console.error("Error saving new unit and chapter:", error);
+      }
```
Comment (alexng353) [inline src/app/admin/subject/[slug]/page.tsx:143]: I feel like this isn't correct

**Review by alexng353 — CHANGES_REQUESTED** [review]

File: src/app/admin/subject/[slug]/[unit]/[articleNumber]/page.tsx (hunk 7 above)
Comment (AlokSwaminathan) [reply src/app/admin/subject/[slug]/[unit]/[articleNumber]/page.tsx:15]: I think this is fine, we have some guarantees about subject names and unit names

File: src/app/admin/subject/[slug]/page.tsx (hunk 11 above)
Comment (AlokSwaminathan) [reply src/app/admin/subject/[slug]/page.tsx:143]: i agree, in the case that the user doesn't have the perms to save a document, you want to give them a clear error, not something in the console which they aren't gonna see
maybe an alert or something would be simplest

Comment (AlokSwaminathan) [reply src/app/admin/subject/[slug]/page.tsx:143]: nvm i see this is an admin gated page that's mb

**Comment by AlokSwaminathan** [comment]
I notice that you are using cookies for auth, there is setCustomUserClaims for firebase which is intended for auth, I think porting to that might be more ideal

**Comment by manningwu07** [comment]
@AlokSwaminathan That was old code; I was trying to use middleware for something but realized a better solution. I deleted that file because it wasnt doing anything.

File: src/app/admin/subject/[slug]/page.tsx (hunk 10 above)
Comment (manningwu07) [reply src/app/admin/subject/[slug]/page.tsx:102]: Nah bruh its a onetime thing. Is it conventional to do that though even though you use it only once?

File: package-lock.json
Comment (alexng353) [inline package-lock.json:1]: Why was this file deleted

File: src/app/admin/page.tsx [hunk 12]
```diff
@@ -1,60 +1,172 @@
...
+                    type="text"
+                    className="mb-3 w-full rounded-md border p-2"
+                    placeholder="Search for a user..."
+                    value={searchTermUsers}
+                    onChange={(e) =>
+                      setSearchTermUsers(e.target.value)
+                    }
+                  />
+                  <ul className="class-list max-h-60 overflow-y-auto">
+                    {error && users!.map((u) => ( // If error, it will show error message. Otherwise, it will show users
```
Comment (alexng353) [inline src/app/admin/page.tsx:71]: This is still so confusing to me, if error is falsy, doesn't that mean that users!.map... will show?

File: src/app/admin/page.tsx (hunk 6 above)
Comment (alexng353) [reply src/app/admin/page.tsx:117]: brug

File: src/app/admin/subject/[slug]/page.tsx [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+          } else {
+            emptyData.title =
+              apClasses.find(
+                (apClass) =>
+                  apClass
+                    .replace(/AP /g, "")
+                    .toLowerCase()
+                    .replace(/[^a-z1-9 ]+/g, "")
+                    .replace(/\s/g, "-") === params.slug,
+              ) || "";
```
Comment (alexng353) [inline src/app/admin/subject/[slug]/page.tsx:52]: What does this do? pls document this because of weird behaviour

File: src/app/article-creator/_components/ArticleCreator.tsx [hunk 14]
```diff
@@ -90,54 +74,58 @@ function ArticleCreator({ className }: { className?: string }) {
...
-  const handleTitleSelect = async (selectedTitle: string) => {
+  const handleTitleSelect = async () => {
     const user = await getUser();
-
-    const formattedTitle = selectedTitle
-      .replace(/AP /g, "")
-      .toLowerCase()
-      .replace(/[^a-z1-9 ]+/g, "")
-      .replace(/\s/g, "-");
+    console.log("user", user);
```
Comment (alexng353) [inline src/app/article-creator/_components/ArticleCreator.tsx:107]: remove conlog

File: src/app/components/hooks/users.ts [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+          }
+        } else {
+          resolve(null);
+        }
+        unsubscribe();
+      },
+      reject,
+    );
+  });
+};
```
Comment (alexng353) [inline src/app/components/hooks/users.ts:46]: try to reduce the level of nesting here by separating parts of the logic out. Also, reverse the conditions so you can use guard clauses

example:
```ts
if (condition) {
} else {
  reject(...)
}

// becomes
if (!condition) {
  reject(...)
}
// original code
...
```

File: src/app/components/hooks/users.ts [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+      await updateDoc(userDocRef, {
+        access: newRole,
+      });
+    } catch (error) {
+      console.error("Error updating user role:", error);
+      throw error;
+    }
+  } else {
+    return;
+  }
```
Comment (alexng353) [inline src/app/components/hooks/users.ts:82]: reverse the condition and early return

File: src/app/components/landingPage/APsection.tsx [hunk 17]
```diff
@@ -33,7 +30,7 @@ const APsection: React.FC<SectionProps> = ({
           {courses.map((course, index) => (
             <li key={index} className="break-inside-avoid-column">
               <Link
-                href={`${domainName}/${currentPath}/subject/${formatCourseName(course)}`}
+                href={`/subject/${formatCourseName(course)}`}
```
Comment (alexng353) [inline src/app/components/landingPage/APsection.tsx:33]: The formatCourseName function actually already exists

File: src/app/components/subjectHomepage/subject-sidebar.tsx [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+  // Handle unit delete
+  const deleteUnit = (unitIndex: number) => {
+    const updatedUnits = [...units];
+    updatedUnits.splice(unitIndex, 1);
+    setUnits(updatedUnits);
+  };
+
+  const pathname = window.location.pathname;
+  const isAdmin = pathname.split("/").includes("admin") ? "/admin" : "";
+  isAdmin && sliceNum++;
```
Comment (alexng353) [inline src/app/components/subjectHomepage/subject-sidebar.tsx:105]: this probably needs to be in a useEffect

File: src/app/components/subjectHomepage/subject-sidebar.tsx [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+
+const SubjectSidebar = (props: Props) => {
+  const [isCollapsed, setIsCollapsed] = useState(false);
+  const [units, setUnits] = useState(props.subject.units);
+  const [editingUnit, setEditingUnit] = useState<number | null>(null);
+  const [editingChapter, setEditingChapter] = useState<{
+    unitIndex: number | null;
+    chapterIndex: number | null;
+  }>({ unitIndex: null, chapterIndex: null });
+  let sliceNum = 3;
```
Comment (alexng353) [inline src/app/components/subjectHomepage/subject-sidebar.tsx:26]: this probably needs to be a state

**Review by alexng353 — CHANGES_REQUESTED** [review]
Why did you move everything into src/app/...

File: src/app/admin/page.tsx (hunk 12 above)
Comment (manningwu07) [reply src/app/admin/page.tsx:71]: I messed up lmfao I changed it dw

File: package-lock.json
Comment (manningwu07) [reply package-lock.json:1]: I thought that was npm 😭

File: src/app/components/landingPage/APsection.tsx (hunk 17 above)
Comment (manningwu07) [reply src/app/components/landingPage/APsection.tsx:33]: That sluggifies. The ones in admin undo this sluggification. I probably need to change it later but for now its fine.

File: src/app/components/subjectHomepage/subject-sidebar.tsx (hunk 19 above)
Comment (manningwu07) [reply src/app/components/subjectHomepage/subject-sidebar.tsx:26]: Set to state

File: src/app/components/subjectHomepage/subject-sidebar.tsx (hunk 18 above)
Comment (manningwu07) [reply src/app/components/subjectHomepage/subject-sidebar.tsx:105]: No cuz then I need to create more "use states"
